### PR TITLE
Router: stability fund swap implementation

### PR DESCRIPTION
**Pull request by TRSSMG**

### CODE REVIEW

**Review by nightowl121 — APPROVED** [review]
Approved

**Review by AntonAndell — CHANGES_REQUESTED** [review]
Route v2 also need to be a option in tokenFallback

File: core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java [hunk 1]
```diff
@@ -276,23 +313,39 @@ public void xTokenFallback(String _from, BigInteger _value, byte[] _data) {
...
-        Address[] path = new Address[pathArray.size()];
-
-        for (int i = 0; i < pathArray.size(); i++) {
-            JsonValue addressJsonValue = pathArray.get(i);
-            if (addressJsonValue == null || addressJsonValue.toString().equals("null")) {
-                path[i] = null;
-            } else {
-                path[i] = Address.fromString(addressJsonValue.asString());
+        List<RouteAction> actions = new ArrayList<>();
+        if (!params.contains("isStableSwap")) {
```
Comment (AntonAndell) [inline core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:317]: Instead use the method field here.
Switch on method and default revert.
"_swap"
and maybe 
"_swapv2"

No need for the extra parameter

File: core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java [hunk 2]
```diff
@@ -276,23 +319,42 @@ public void xTokenFallback(String _from, BigInteger _value, byte[] _data) {
...
         }
 
         Address fromToken = Context.getCaller();
-        route(receiver, fromToken, path, minimumReceive);
+        route(receiver, fromToken, actions, minimumReceive);
+    }
+
+    private void tokenFallbackV2(String _from, byte[] data) {
+        RouteData routeData = RouteData.fromBytes(data);
+        Context.require(routeData.method.contains("_swapV2"), TAG + ": Fallback directly not allowed.");
```
Comment (AntonAndell) [inline core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:342]: Do we need a method at all here?

**Review by AntonAndell — COMMENTED** [review]
Looks good, just one minor comment for now. WIll take another look tomorrow

File: core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java (hunk 2 above)
Comment (TRSSMG) [reply core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:342]: ya, I think, separation is better

Comment (AntonAndell) [reply core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:342]: How do you mean separation? What does the extra string "_swapv2" add information wise?

I can perhaps see if we want to add more methods in the future. But we already have actions for that

File: core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java [hunk 3]
```diff
@@ -283,17 +287,41 @@ public void tokenFallback(Address _from, BigInteger _value, byte[] _data) {
 
     @External
     public void xTokenFallback(String _from, BigInteger _value, byte[] _data) {
-        // Receive token transfers from Balanced DEX and staking while in mid-route
-        if (_from.equals(getDex().toString()) || _from.equals(MINT_ADDRESS.toString())) {
+        if (inRoute) {
             return;
```
Comment (TRSSMG) [inline core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:291]: this also should be revert, other wise the token transfer will be success, while underlying reason for transfer will not execute

Comment (AntonAndell) [reply core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:291]: No, this we need to accept deposits during the route. Only if someone tries to reenter(which they should not be able to) the will simply lose their deposit

Comment (TRSSMG) [reply core-contracts/Router/src/main/java/network/balanced/score/core/router/RouterImpl.java:291]: ok, got it

**Review by AntonAndell — APPROVED** [review]

**Pull request merged**